src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs

```diff
@@ -38,10 +120,169 @@ private void SetContentDispositionHeader(ActionContext context, FileResult resul
+            for (var i = 0; i < states.Length; i++)
+            {
+                if (states[i] > max)
+                {
+                    max = states[i];
+                }
+            }
+            return max;
+        }
+
+        private static (RangeItemHeaderValue range, long rangeLength) SetRangeHeaders(
+            ActionContext context,
+            RequestHeaders httpRequestHeaders,
+            long? fileLength,
+            DateTimeOffset? lastModified = null,
+            EntityTagHeaderValue etag = null)
+        {
+            var response = context.HttpContext.Response;
+            var httpResponseHeaders = response.GetTypedHeaders();
+            var range = fileLength == null ? null : ParseRange(context, httpRequestHeaders, fileLength.Value, lastModified, etag);
+            var rangeNotSatisfiable = range == null;
+            if (rangeNotSatisfiable)
+            {
+                // 14.16 Content-Range - A server sending a response with status code 416 (Requested range not satisfiable)
+                // SHOULD include a Content-Range field with a byte-range-resp-spec of "*". The instance-length specifies
+                // the current length of the selected resource.  e.g. */length
+                response.StatusCode = StatusCodes.Status416RangeNotSatisfiable;
+                if (fileLength.HasValue)
+                {
+                    httpResponseHeaders.ContentRange = new ContentRangeHeaderValue(fileLength.Value);
+                }
+                return (null, fileLength.Value);
+            }
+
+            httpResponseHeaders.ContentRange = new ContentRangeHeaderValue(
+                range.From.Value,
+                range.To.Value,
+                fileLength.Value);
+
+            response.StatusCode = StatusCodes.Status206PartialContent;
+            var rangeLength = SetContentLength(context, range);
+            return (range, rangeLength);
+        }
+
+        private static long SetContentLength(ActionContext context, RangeItemHeaderValue range)
+        {
+            var start = range.From.Value;
+            var end = range.To.Value;
+            var length = end - start + 1;
+            var response = context.HttpContext.Response;
+            response.ContentLength = length;
+            return length;
+        }
+
+        private static RangeItemHeaderValue ParseRange(
+            ActionContext context,
+            RequestHeaders httpRequestHeaders,
+            long fileLength,
+            DateTimeOffset? lastModified = null,
+            EntityTagHeaderValue etag = null)
+        {
+            var httpContext = context.HttpContext;
+            var response = httpContext.Response;
+
+            var range = RangeHelper.ParseRange(
+                httpContext,
+                httpRequestHeaders,
+                lastModified,
+                etag);
+
+            if (range != null)
+            {
+                var rangeValue = range.Single();
```

Comment:
That's how the helper is set up. RangeHelper.ParseRange returns a collection of Ranges from the range header. Should we revisit the helper?